File: task/src/main/java/com/acutecoder/task/Task.java
package com.acutecoder.task;

import android.os.Handler;
import android.os.Looper;

/**
 * Runs a task in Background thread or Foreground thread (UI thread)
 * Created by Bhuvaneshwaran
 * on 11:00 PM, 4/10/2023
 *
 * @author AcuteCoder
 */

public class Task<T> {

    private final TaskRunnable<T> taskRunnable;
    private TaskResult<T> taskResult;
    private TaskError taskError;
    private TaskCallback onStart, onEnd;
    private boolean isBackground = true, isRunning = false;
    private OnProgress onProgress;
    private NextTask<T> next;

    /**
     * Creates a new Task instance
     *
     * @param taskRunnable The task
     */
    public Task(TaskRunnable<T> taskRunnable) {
        this.taskRunnable = taskRunnable;
    }

    /**
     * Creates a new Task instance
     *
     * @param taskRunnable The task
     * @return Task instance
     */
    public static Task<?> with(TaskRunnable<?> taskRunnable) {
        return new Task<>(taskRunnable);
    }

    private static Handler getForegroundHandler() {
        return new Handler(Looper.getMainLooper());
    }

    /**
     * Called when the task is successfully completed without error
     */
    public Task<T> onResult(TaskResult<T> taskResult) {
        this.taskResult = taskResult;
        return this;
    }

    /**
     * Called when the task is failed
     */
    public Task<T> onError(TaskError taskError) {
        this.taskError = taskError;
        return this;
    }

    /**
     * Called before the execution of task
     */
    public Task<T> onStart(TaskCallback onStart) {
        this.onStart = onStart;
        return this;
    }

    /**
     * Called after the execution of task regardless the completion of the task (Whether the task is executed without error)
     */
    public Task<T> onEnd(TaskCallback onEnd) {
        this.onEnd = onEnd;
        return this;
    }

    /**
     * Called to update progress (in UI) by publishProgress function in UI Thread
     */
    public Task<T> onProgress(OnProgress onProgress) {
        this.onProgress = onProgress;
        return this;
    }

    /**
     * Runs the task in background thread
     */
    public Task<T> doInBackground() {
        isBackground = true;
        return this;
    }

    /**
     * Runs the task in foreground thread (Main thread or UI thread)
     */
    public Task<T> doInForeground() {
        isBackground = false;
        return this;
    }

    /**
     * Starts the execution of the task
     */
    public void start() {
        try {
            if (isRunning) throw new TaskException("Task already running!", isBackground);
            isRunning = true;
            if (isBackground) {
                try {
                    if (onStart != null) onStart.run();
                    new Thread(new java.lang.Runnable() {
                        @Override
                        public void run() {
                            try {
                                assert taskRunnable != null;
                                final T o = taskRunnable.run(Task.this);
                                getForegroundHandler().post(new java.lang.Runnable() {
                                    @Override
                                    public void run() {
                                        if (taskResult != null) taskResult.onResult(o);
                                        if (onEnd != null) onEnd.run();
                                        if (next != null) next.run(o).start();
                                    }
                                });
                                isRunning = false;
                            } catch (Exception e) {
                                isRunning = false;
                                if (onEnd != null)
                                    getForegroundHandler().post(new java.lang.Runnable() {
                                        @Override
                                        public void run() {
                                            onEnd.run();
                                        }
                                    });
                                if (taskError != null) Task.this.onError(e);
                                else throw new TaskException(e, isBackground);
                            }
                        }
                    }).start();
                } catch (Exception e) {
                    if (taskError != null) onError(e);
                    else throw new TaskException(e, isBackground);
                }
            } else {
                try {
                    getForegroundHandler().post(new java.lang.Runnable() {
                        @Override
                        public void run() {
                            try {
                                if (onStart != null) onStart.run();
                                T o = taskRunnable.run(Task.this);
                                if (taskResult != null) {
                                    taskResult.onResult(o);
                                }
                                isRunning = false;
                                if (onEnd != null) onEnd.run();
                                if (next != null) next.run(o).start();
                            } catch (Exception e) {
                                isRunning = false;
                                if (onEnd != null) onEnd.run();
                                if (taskError != null) Task.this.onError(e);
                                else throw new TaskException(e, isBackground);
                            }
                        }
                    });
                } catch (Exception e) {
                    if (taskError != null) onError(e);
                    else throw new TaskException(e, isBackground);
                }
            }
        } catch (Exception e) {
            if (taskError != null) onError(e);
            else throw new TaskException(e, isBackground);
        }
    }

    /**
     * Publish the progress in UI Thread
     */
    public Task<T> publishProgress(final Object... progress) {
        if (onProgress != null)
            getForegroundHandler().post(new Runnable() {
                @Override
                public void run() {
                    onProgress.progress(progress);
                }
            });
        return this;
    }

    /**
     * Sleeps thread for given milliseconds
     */
    public Task<T> sleep(long mills) throws InterruptedException {
        Thread.sleep(mills);
        return this;
    }

    /**
     * Chains next Task
     */
    public Task<T> then(NextTask<T> nextTask) {
        next = nextTask;
        return this;
    }

    private void onError(final Exception e) {
        getForegroundHandler().post(new java.lang.Runnable() {
            @Override
            public void run() {
                taskError.onError(e);
            }
        });
    }

    public static class Foreground {
        /**
         * Runs a simple task in UI thread
         */
        public static void start(java.lang.Runnable runnable) {
            getForegroundHandler().post(runnable);
        }
    }

    public static class Background {
        /**
         * Runs a simple task in new thread
         */
        public static void start(java.lang.Runnable runnable) {
            new Thread(runnable).start();
        }
    }
}

File: taskkt/src/main/java/com/acutecoder/task/Task.kt
package com.acutecoder.task

import android.os.Handler
import android.os.Looper

/**
 * Runs a task in Background thread or Foreground thread (UI thread)
 *
 * Created by Bhuvaneshwaran
 *
 * 4:30 PM, 8/28/2023
 *
 * @author AcuteCoder
 */

class Task<T>(private val runnable: Task<T>.() -> T) : AbstractTask<T> {
    private var taskResult: ((result: T) -> Unit)? = null
    private var taskError: ((e: Exception) -> Unit)? = null
    private var onStart: (() -> Unit)? = null
    private var onEnd: (() -> Unit)? = null
    private var onCancel: (() -> Unit)? = null
    private var onProgress: ((Array<out Any>) -> Unit)? = null
    private var isBackground = true
    private var isRunning = false
    private var isCancelled = false
    private var next: ((T) -> AbstractTask<*>)? = null

    /**
     * Called when the task is successfully completed without error
     */
    override fun onResult(taskResult: ((result: T) -> Unit)?): Task<T> {
        this.taskResult = taskResult
        return this
    }

    /**
     * Called when the task is failed
     */
    override fun onError(taskError: ((error: Exception) -> Unit)?): Task<T> {
        this.taskError = taskError
        return this
    }

    /**
     * Called before the execution of task
     */
    override fun onStart(onStart: (() -> Unit)?): Task<T> {
        this.onStart = onStart
        return this
    }

    /**
     * Called after the execution of task regardless the completion of the task (Whether the task is executed without error)
     */
    override fun onEnd(onEnd: (() -> Unit)?): Task<T> {
        this.onEnd = onEnd
        return this
    }

    /**
     * Called when the execution of task is cancelled
     */
    override fun onCancel(onCancel: (() -> Unit)?): Task<T> {
        this.onCancel = onCancel
        return this
    }

    /**
     * Called to update progress (in UI) by publishProgress function in UI Thread
     */
    override fun onProgress(onProgress: ((Array<out Any>) -> Unit)?): Task<T> {
        this.onProgress = onProgress
        return this
    }

    /**
     * Runs the task in background thread
     */
    fun doInBackground(): Task<T> {
        isBackground = true
        return this
    }

    /**
     * Runs the task in foreground thread (Main thread or UI thread)
     */
    fun doInForeground(): Task<T> {
        isBackground = false
        return this
    }

    /**
     * Starts the execution of the task
     */
    override fun start() {
        try {
            isCancelled = false
            if (isRunning) throw TaskException("Task already running!", isBackground)
            isRunning = true
            if (isBackground) {
                try {
                    onStart?.invoke()
                    Thread {
                        try {
                            val o = runnable(this@Task)
                            foregroundHandler.post {
                                taskResult?.invoke(o)
                                onEnd?.invoke()
                                next?.invoke(o)?.start()
                            }
                            isRunning = false
                        } catch (e: CancellationException) {
                            isRunning = false;
                            onCancel?.let {
                                foregroundHandler.post {
                                    it.invoke()
                                }
                            }
                        } catch (e: Exception) {
                            isRunning = false
                            onEnd?.let { foregroundHandler.post { it.invoke() } }
                            if (taskError != null)
                                onError(e)
                            else
                                throw TaskException(e, isBackground)
                        }
                    }.start()
                } catch (e: Exception) {
                    if (taskError != null) onError(e) else throw TaskException(e, isBackground)
                }
            } else {
                try {
                    foregroundHandler.post {
                        try {
                            onStart?.invoke()
                            val o = runnable(this@Task)
                            taskResult?.invoke(o)
                            isRunning = false
                            onEnd?.invoke()
                            next?.invoke(o)?.start()
                        } catch (e: CancellationException) {
                            isRunning = false;
                            onCancel?.invoke()
                        } catch (e: Exception) {
                            isRunning = false
                            onEnd?.invoke()
                            if (taskError != null)
                                onError(e)
                            else
                                throw TaskException(e, isBackground)
                        }
                    }
                } catch (e: Exception) {
                    if (taskError != null) onError(e) else throw TaskException(e, isBackground)
                }
            }
        } catch (e: Exception) {
            if (taskError != null) onError(e) else throw TaskException(e, isBackground)
        }
    }

    /**
     * Publish the progress in UI Thread
     */
    override fun publishProgress(vararg progress: Any): Task<T> {
        onProgress?.let {
            foregroundHandler.post {
                it.invoke(progress)
            }
        }
        return this
    }

    /**
     * Sleeps thread for given milliseconds
     */
    fun sleep(mills: Long): Task<T> {
        Thread.sleep(mills)
        return this
    }

    /**
     * Chains next Task
     */
    override fun then(nextTask: ((T) -> AbstractTask<*>)?): Task<T> {
        next = nextTask;
        return this
    }

    /**
     * Sets the cancellation flag
     */
    override fun cancel(): Task<T> {
        isCancelled = true
        return this
    }

    /**
     * Returns state of task
     */
    override  fun isActive(): Boolean {
        return !isCancelled
    }

    /**
     * Ensure task is alive (ie, task is not cancelled)
     */
    override fun ensureActive() {
        if (!isActive()) throw CancellationException()
    }

    private fun onError(e: Exception) {
        foregroundHandler.post { taskError?.invoke(e) }
    }

    object Foreground {
        /**
         * Runs a simple task in UI thread
         */
        fun start(runnable: Runnable?) {
            foregroundHandler.post((runnable)!!)
        }
    }

    object Background {
        /**
         * Runs a simple task in new thread
         */
        fun start(runnable: Runnable?) {
            Thread(runnable).start()
        }
    }

    companion object {
        private val foregroundHandler = Handler(Looper.getMainLooper())
    }
}